File: paleo_workbench/ui/deferred_page_bindings.py
"""Small main-thread queue for non-visible project-page bindings.

Project open needs the data/navigation page immediately, while scientific
pages may load previews or native state as part of ``set_project`` /
``update_state``.  This queue holds only the newest callback for each named
binding and intentionally owns no project model: callers retain authority for
the live document and invoke :meth:`flush` on the GUI thread.
"""

from __future__ import annotations

from collections.abc import Callable


class DeferredPageBindings:
    """Coalesce newest non-visible page bindings until first visit."""

    def __init__(self) -> None:
        self._pending: dict[int, dict[str, Callable[[], None]]] = {}

    def schedule(self, index: int, name: str, callback: Callable[[], None]) -> None:
        self._pending.setdefault(index, {})[name] = callback

    def flush(self, index: int) -> None:
        # A project binding may enqueue the matching state callback.  Drain
        # until stable so first navigation never requires a second click; run
        # project assignment ahead of state regardless of enqueue order.
        while callbacks := self._pending.pop(index, None):
            project = callbacks.pop("project", None)
            if project is not None:
                project()
            for callback in callbacks.values():
                callback()

    def has_pending(self, index: int) -> bool:
        return bool(self._pending.get(index))

File: libs/ui_shell/src/deferred_page_bindings.cpp
#include "pwb/ui_shell/deferred_page_bindings.hpp"

#include <utility>

namespace pwb::ui_shell {

void DeferredPageBindings::schedule(int index, std::string name,
                                    std::function<void()> callback) {
    auto& list = pending_[index];
    for (auto& entry : list) {
        if (entry.first == name) {
            entry.second = std::move(callback);
            return;
        }
    }
    list.emplace_back(std::move(name), std::move(callback));
}

void DeferredPageBindings::flush(int index) {
    // Drain until stable: bindings may enqueue further bindings for the same
    // index during the flush.
    for (;;) {
        const auto it = pending_.find(index);
        if (it == pending_.end() || it->second.empty()) {
            return;
        }
        BindingList callbacks = std::move(it->second);
        pending_.erase(it);

        // "project" assignment runs ahead of state callbacks regardless of
        // enqueue order.
        std::function<void()> project;
        BindingList rest;
        rest.reserve(callbacks.size());
        for (auto& [name, cb] : callbacks) {
            if (name == "project" && !project) {
                project = std::move(cb);
            } else {
                rest.emplace_back(std::move(name), std::move(cb));
            }
        }
        if (project) {
            project();
        }
        for (auto& [name, cb] : rest) {
            if (cb) {
                cb();
            }
        }
    }
}

bool DeferredPageBindings::has_pending(int index) const {
    const auto it = pending_.find(index);
    return it != pending_.end() && !it->second.empty();
}

}  // namespace pwb::ui_shell
